File: dynamic/15/test1.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {
	List<List<Integer>> result;
	List<Integer> solu;

	public List<List<Integer>> combinationSum(int[] candidates, int target) {
		result = new ArrayList();
		solu = new ArrayList();
		Arrays.sort(candidates);
		getCombination(candidates, target, 0, 0);
		return result;
	}

	public void getCombination(int[] candidates, int target, int sum, int level) {
		if(sum>target)
			return ;
		if (solu.size()==2) {
			if(Arrays.binarySearch(candidates,level,candidates.length,(target-sum))>=0){
				solu.add(target-sum) ;
				result.add(new ArrayList(solu));
				solu.remove(solu.size()-1);
			}
			return;
		}
		for (int i = level; i < candidates.length; i++) {
			sum += candidates[i];
			solu.add(candidates[i]);
			getCombination(candidates, target, sum, i + 1);
			solu.remove(solu.size() - 1);
			sum -= candidates[i];
			while(i+1 < candidates.length&&candidates[i]==candidates[i+1]){
				i++;
			}
			
		}
	}
	
	public static void main(String args[]){
		Solution s = new Solution() ;
		int[] candidates = {-1,0,1,2,-1,-4};
		Arrays.sort(candidates);
//		int index = Arrays.binarySearch(candidates,4,candidates.length,1) ;
//		System.out.println(index);
		List<List<Integer>> list = s.combinationSum(candidates, 0) ;
		for(List<Integer> l : list){
			for(Integer i : l){
				System.out.print(i+" ");
			}
			System.out.println();
		}
	}
}

File: dynamic/39/test1.c
#define MAX 100
#include<stdio.h>
int getRow(int* candidates,int candidatesSize,int target){
	int i=0 ;
        for(i=0;i<candidatesSize;i++){
                if(candidates[i]>target){
                        break ;
                }
        }
        return i ;
}
int combinationSum(int* candidates, int candidatesSize, int target){
	int row =getRow(candidates,candidatesSize,target) ;
        int m[MAX][MAX] ;
        for(int i=0;i<row;i++){
                m[i][0] = 0 ;
        }
        for(int j=1;j<=target;j++){
                for(int i=0;i<row;i++){
                                int sum = 0 ;
                                if(j%candidates[i]==0){
                                     sum = 1 ;
                                }else{

                                }
                                if(i>0){
                                        int bound = j / candidates[i] ;
                                        for(int n=0;n<=bound;n++){
                                                sum =sum + m[i-1][j-n*candidates[i]] ;
                                        }
                                }
                                m[i][j] = sum ;
                }
        }
	return m[row-1][target] ;

}

int main(){
	int candidates[] = {8,10,6,3,4,12,11,5,9} ;
	int candidatesSize = 9;
	int target =28  ;
	int result = combinationSum(candidates,candidatesSize,target) ;
	printf("%d\n",result) ;
	return 0 ;
}
